RX - Corrections to RX fail detection, hold and preset, when using a PWM connection

ProDrone opened the pull request:
@hydra This is an important update to #995.

RX hold values were not preserved and FC preset values were only applied after an initial jump to the RX failsafe preset value under the following conditions:

Use PWM. Set channels 1-4 (sticks) to failsafe OFF. Set aux channel to a preset failsafe value.

This PR has been tested with PWM, PPM and SUMD (using GR-12), in all cases FC hold function `rxfail 0 h 1600`and FC preset function `rxfail 0 s 1600` worked well.

More has changes, please see commit message for details.

borisbstyle commented:
This looks good to me.

These min_usec max_sec are now being checked multiple times in the code from sampling to detecting failsafe etc.
Perhaps that can be optimized now since you are doing it already in early stage?

ProDrone commented:
@borisbstyle The first check in 
~~~
rxUpdateFlightChannelStatus(channel, sample);
~~~
is for validation of the first four the control channels and is also used for unittests.
The second one in:
~~~
if (sample < rxConfig->rx_min_usec || sample > rxConfig->rx_max_usec || !rxSignalReceived) {
    sample = getRxfailValue(channel);
}
~~~
is to check and correct individual (aux) channels. The only optimization i see is that this check can be skipped for the (4) control channels. That provides a tiny gain (if any) in execution time and adds code for the check (oh, and yes, we should forget about coding one of these days and fly instead :smile:

ProDrone commented:
@hydra What are your thoughts about this commit https://github.com/ProDrone/cleanflight/commit/dd4b6b6286fdd5502eee97158071bec1a1ee5d4a ?

ProDrone commented:
@hydra I removed some commits and moved them to their own PR: #1194 and #1195
__*DO NOT YET MERGE !*__ (I found erratic behavior using PPM - investigating :mag: )

hydra commented:
What's the status of this ProDrone.

I would like to get 1.10.0 released and this is blocking the release.

ProDrone commented:
@hydra Not solved yet. Now that i know it is blocking and marked critical, i will focus on this one exclusively. Looking into it right now... I expect a solution tomorrow.

ProDrone commented:
@hydra :: __*READY FOR MERGE*__

hydra commented:
@ProDrone first step - firmware. Second step UI.

borisbstyle commented:
Good work @ProDrone !

ProDrone commented:
Thanks! @borisbstyle. Still not completely happy with how it is. (know the feeling? :grinning:)

@hydra I would prefer to add the the first four (control) channels also to the rxfail command:
Where `rxfail 0 h 1500` will hold the last value until failsafe handling kicks in; and
Where `rxfail 0 s 1500` uses the current behavior (set as a default for safety).

ProDrone commented:
@hydra Can you please give your thoughts on this one?
If you like this addition,i will make a PR for it.

borisbstyle commented:
I agree. A glitch should not abruptly go down to min_usec. It should be
configurable.
Op 12 aug. 2015 17:33 schreef "ProDrone" <notifications@github.com>:

> @hydra <https://github.com/hydra> Can you please give your thoughts on
> this one?
> If you like this addition,i will make a PR for it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1190#issuecomment-130344141>
> .
>

ProDrone commented:
@borisbstyle Thats the other path i would like to modify... Make them going a bit smoother moving to their targets...

ProDrone commented:
@hydra Please also merge #1203 to 1.10.0 release, it will make a lot of people happy i think...
__*When is your deadline for 1.10.0 ?*__ (To see if there is still time for another feature i like to add: make the RPYT channels moving smoother to their targets...)

borisbstyle commented:
@prodrone @hydra
Since I rebased with the latest master I have some PWM users reporting issues. There is erratic behaviour where it keeps jumping between preset value and value from rx. The problems were not there before these latest prs
I will post some detaills later.

jaapp commented:
@ProDrone Thanks for your work!

borisbstyle commented:
There is a guy with Spectrum AR5000 who gets a pulsing motors on PWM with the latest master. Constant failsafe. It is like failsafe thinks that it's failsafe (false detection)

He is trying to get me the blackbox logs.

borisbstyle commented:
This is the logfile
https://www.dropbox.com/s/f6kmqrg907atrsw/LOG00078.TXT?dl=0

I am now busy with removing this PR for him and see how it goes after that

borisbstyle commented:
And without this PR it works like expected!

@hydra 
This PR needs to be reverted or fixed. Don't release 1.10 yet!

ProDrone commented:
Failsafe detection depends for a big part on correct configuration of the receiver, especially when using PWM. The problems describe here can easily be caused by a wrong configuration.

The provided CLI dump is useful, the blackbox log shows the problem is real, but does not provide a clue about the cause. I have not seen any info about the receivers failsafe settings.

I like to see if i can reproduce the problem.

@borisbstyle Is it correct that the faulty one is 5062ef10af7f2ebfbbe56157af02bc4017257b06 or are the published .hex file the result of local code or a different branch?

borisbstyle commented:
BTW I created an issue for this for further discussing.
https://github.com/cleanflight/cleanflight/issues/1207

No I rebased my branch in between and left this one out....i changed the history.
The hex he used was the real master (not betaflight). The issue popped up when I rebased betaflight to the master.
I wanted to make sure it was really something in the master for his case and let him try with master and same thing. Removed this PR and problem was gone.

I am not sure what is causing it in his case, but this is certainly something blocking as you cannot expect everyone to reconfigure their RX after update. You can also expect different behaviour with different brands and types on PWM.
I am not able to help further as I don't have may different RX's to help with this PR. I am not familliar with different behaviours.
I did ask him to look in the issue I created and provide more information there about his config.

ProDrone commented:
Since you opened an issue #1207 for this, i will comment there...

ProDrone commented:
@hydra __*Please do not revert the complete PR*__.
I am working on a fix to apply on top of this, to remove the troubled part but leave good changes in place. I have it already, locally tested too, but want some feedback before making a PR...
Discussion has been diverted to #1207
